smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/httpResponse/CodedErrorGenerator.kt

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+package software.amazon.smithy.swift.codegen.integration.httpResponse
+
+import software.amazon.smithy.aws.traits.protocols.AwsQueryErrorTrait
+import software.amazon.smithy.codegen.core.Symbol
+import software.amazon.smithy.model.shapes.OperationShape
+import software.amazon.smithy.model.shapes.StructureShape
+import software.amazon.smithy.swift.codegen.SwiftDependency
+import software.amazon.smithy.swift.codegen.declareSection
+import software.amazon.smithy.swift.codegen.integration.ProtocolGenerator
+import software.amazon.smithy.swift.codegen.integration.SectionId
+import software.amazon.smithy.swift.codegen.integration.middlewares.handlers.MiddlewareShapeUtils
+import software.amazon.smithy.swift.codegen.model.getTrait
+
+class CodedErrorGenerator(
+    val ctx: ProtocolGenerator.GenerationContext,
+    val op: OperationShape,
+    val unknownServiceErrorSymbol: Symbol
+) {
+    object CodedErrorGeneratorSectionId : SectionId
+
+    fun render() {
+        val errorShapes = op.errors.map { ctx.model.expectShape(it) as StructureShape }.toSet().sorted()
+        val operationErrorName = MiddlewareShapeUtils.outputErrorSymbolName(op)
+        val rootNamespace = ctx.settings.moduleName
+        val httpBindingSymbol = Symbol.builder()
+            .definitionFile("./$rootNamespace/models/$operationErrorName+CodedError.swift")
+            .name(operationErrorName)
+            .build()
+
+        ctx.delegator.useShapeWriter(httpBindingSymbol) { writer ->
+            writer.addImport(SwiftDependency.CLIENT_RUNTIME.target)
+            writer.addImport(unknownServiceErrorSymbol)
+            val unknownServiceErrorType = unknownServiceErrorSymbol.name
+
+            val context = mapOf(
+                "ctx" to ctx,
+                "unknownServiceErrorType" to unknownServiceErrorType,
+                "operationErrorName" to operationErrorName,
+                "errorShapes" to errorShapes
+            )
+            writer.declareSection(CodedErrorGeneratorSectionId, context) {
+                writer.openBlock("extension \$L: CodedError {", "}", operationErrorName) {
+                    writer.write("")
+                    writer.openBlock("public var errorCode: String? {", "}") {
+                        writer.write("switch self {")
+                        for (errorShape in errorShapes) {
+                            var errorShapeName = resolveErrorShapeName(errorShape)
+                            var errorShapeType = ctx.symbolProvider.toSymbol(errorShape).name
+                            var errorShapeEnumCase = errorShapeType.decapitalize()
+                            writer.write("case .\$L: return \$S", errorShapeEnumCase, errorShapeName)
```

Comment:
Why do we return the error type as a string here? Should we return nil instead?

Reply:
Returning the error type as a string here prevents me from having to extend every error on the operation with its own `CodedError` compliance... the `GetWidgetOutputError` knows the right error code for every possible error it can return.

Compare this to the output error initializer which switches over the expected values of errorCode to determine what exact error to create for the operation & this method makes more sense.  Here's an (abridged) initializer for an operation from an AWS service:
```swift
extension ModifyDBInstanceOutputError {
    public init(errorType: Swift.String?, httpResponse: ClientRuntime.HttpResponse, decoder: ClientRuntime.ResponseDecoder? = nil, message: Swift.String? = nil, requestID: Swift.String? = nil) throws {
        switch errorType {
        case "AuthorizationNotFound" : self = .authorizationNotFoundFault(try AuthorizationNotFoundFault(httpResponse: httpResponse, decoder: decoder, message: message, requestID: requestID))
        case "CertificateNotFound" : self = .certificateNotFoundFault(try CertificateNotFoundFault(httpResponse: httpResponse, decoder: decoder, message: message, requestID: requestID))
        default : self = .unknown(UnknownAWSHttpServiceError(httpResponse: httpResponse, message: message, requestID: requestID))
        }
    }
}
```

`AuthorizationNotFoundFault` and `CertificateNotFoundFault` don't need to be `CodedError`s because `ModifyDBInstanceOutputError` returns the right error code for them.

Reply:
The extension I created is basically this switch statement in reverse... convert the enum case to a string instead of convert the string to an enum case.